### DISCUSSION TITLE
Help me understand Emgu CV

### DISCUSSION THREAD

**Discussion by profdeveloper231** [emgucv/emgucv#823] [Q&A]
So I know there are two versions of emgu cv for a developer writing an image processing solution targeting the windows OS and they are :
1. Emgu CV core library
2. Emgu CV run time

Now from my experience installing just the core library into a solution throws an exception that is related to initializing the instance of Emgu Cv in a sample app. Both the library and the runtime are needed to call the methods and functions during development. So my question is, is Emgu Cv like Tesseract where the target OS must have the Emgu CV SDK or runtime installed and the app relying on this runtime must have the emgu cv core library installed to work? Thank You for your answers. It's because during development I have the runtime nuget package installed and it works for my application but when I generate a setup project and install on my PC it throws an exception that is related to the runtime not being found. How can I make the setup download and install the required runtime during installation, Thank You.

**Comment by emgucv** [comment]
1. Check the build folder during developement, it contains all files need for Emgu CV to run. 
2. If you use package installer, check the file contents after the install, they should install all files similar to the development folder. 
3. Follow this steps to use dependency walker to open "cvextern.dll", check if there are missing dependencies. https://www.emgu.com/wiki/index.php/Download_And_Installation#Are_you_missing_any_dependency.3F

**Reply by profdeveloper231** [reply]
Hello thanks like I have no build folder on my project only Debug and Release. I saw something related to setting the configuration to Any CPU will cause neither of the runtime to be copied to the output folder but setting it to either x86 or 64 will make the setup copy the files during deployment but it doesn't still work. I guess if you could explain to me how to add the emgu cv dependency to a project from scratch could help me. Why do we need both the library and the runtime? ,  Thank You

**Comment by emgucv** [comment]
What version of Emgu CV are you using?

Are you using .Net Framework? If you are, this provide the full explanation:
https://www.emgu.com/wiki/index.php/Download_And_Installation#Targeting_.Net_Framework
You cannot use "Any CPU" with .Net Framework, you have to set it to either "x86" or "x64". 

If you are using Emgu CV v4.6 and .Net Core or .Net 5 or above, setting the configuration to "Any CPU" will deploy all 3 native runtimes (win-x86, win-x64, win-arm64) to "runtimes\" folder. 

The Emgu.CV nuget pakcage contains the managed .Net layer (C#) that communicates with the native open cv library. The Emgu.CV.runtime.{} nuget package contains the Open CV (C++) components.

This page contains instructions for adding Emgu CV to your project:
https://www.emgu.com/wiki/index.php/Download_And_Installation#Nuget

**Comment by profdeveloper231** [comment]
Thanks for your effort and contribution. I am using Emgu CV version 4.4.x.x for both the runtime and the core library. Both are installed in my project though I installed the core library later because the project used to work fine with the only the runtime installed. The project works in development environment but throws an exception when used from an installed application. Does Emgu Cv require some kind of license for its software and could this exception be a license issue.

**Reply by ephr4321** [reply]
No, it's not a license issue, since EMGU is GPL [or commercial], and since it cannot test your app's license and know if your app is GPL too - it should just run.
Did you manage making it work?

**Comment by profdeveloper231** [comment]
No it does not work, it still throws the exception "The type CvInvoke threw an exception" from the deployed setup but works when debugged from Visual Studio

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by emgucv** [reply]
Are you using .Net Framework? .Net 6? .Net 7? It was ask before but we did not get an answer. 

Are you using the nuget package from nuget.org? Is the following package used?
https://www.nuget.org/packages/Emgu.CV.runtime.windows
Are you using [4.5.4.4788](https://www.nuget.org/packages/Emgu.CV.runtime.windows/4.5.4.4788) release?

In my earlier post I asked:
  

> Follow this steps to use dependency walker to open "cvextern.dll", check if there are missing dependencies. https://www.emgu.com/wiki/index.php/Download_And_Installation#Are_you_missing_any_dependency.3F

Can you upload the  ".dwi" file output from the dependency walker on the system that throws the exception?

**Reply by profdeveloper231** [reply]
I am using the .NET framework and yes both the runtime and the library are installed. No the version am using 4.4.x.x but I will confirm the exact version and update this post.
